File: content/search/source/includes/facet-tutorial/code-snippets/kotlin-coroutine/tutorial-variable.kt
import com.mongodb.kotlin.client.coroutine.MongoClient
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.runBlocking
import org.bson.Document
import java.time.Instant
import java.util.Date

fun main() {
    // Connection URI for your MongoDB Atlas cluster
    val uri = "<connection-string>"
    val mongoClient = MongoClient.create(uri)

    // Set namespace (database and collection)
    val database = mongoClient.getDatabase("sample_mflix")
    val collection = database.getCollection<Document>("movies")

    runBlocking {
        // Define the aggregation pipeline
        val searchStage = Document("\$search", Document("index", "facet-tutorial")
            .append("facet", Document("operator", Document("near", Document("path", "released")
                // Replace +00:00 with Z
                .append("origin", Date.from(Instant.parse("1921-11-01T00:00:00.000Z"))) 
                .append("pivot", 7776000000L)))
                .append("facets", Document("genresFacet", Document("type", "string")
                    .append("path", "genres"))
                    .append("yearFacet", Document("type", "number").append("path", "year")
                        .append("boundaries", listOf(1910, 1920, 1930, 1940))))))

        val facetStage = Document("\$facet", Document("meta", listOf(
            Document("\$replaceWith", "\$\$SEARCH_META"),
            Document("\$limit", 1)
        )))

        val setStage = Document("\$set", Document("meta", Document("\$arrayElemAt", listOf("\$meta", 0))))

        // Run the aggregation pipeline and print results
        val resultsFlow = collection.aggregate<Document>(
            listOf(searchStage, facetStage, setStage)
        )

        resultsFlow.collect { println(it.toJson()) } // Convert each result to JSON and print
    }

    // Close the MongoDB client
    mongoClient.close()
}

File: content/search/source/includes/facet-tutorial/code-snippets/go/tutorial-variable.go
package main

import (
	"context"
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

func main() {
	// Connects to your Atlas cluster
	client, err := mongo.Connect(options.Client().ApplyURI("<connection-string>"))
	if err != nil {
		panic(err)
	}
	defer client.Disconnect(context.TODO())

	// Sets the namespace
	collection := client.Database("sample_mflix").Collection("movies")

	// Defines the pipeline stages
	searchStage := bson.D{{Key: "$search", Value: bson.M{
		"index": "facet-tutorial",
		"facet": bson.M{
			"operator": bson.M{
				"near": bson.M{
					"path":   "released",
					"origin": time.Date(1921, time.November, 1, 0, 0, 0, 0, time.UTC),
					"pivot":  7776000000},
			},
			"facets": bson.M{
				"genresFacet": bson.M{
					"path": "genres",
					"type": "string"},
				"yearFacet": bson.M{
					"path":       "year",
					"type":       "number",
					"boundaries": bson.A{1910, 1920, 1930, 1940}},
			},
		},
	}}}

	facetStage := bson.D{{Key: "$facet", Value: bson.D{
		{Key: "meta", Value: bson.A{
			bson.D{{Key: "$replaceWith", Value: "$$SEARCH_META"}},
			bson.D{{Key: "$limit", Value: 1}},
		}},
	}}}
	setStage := bson.D{{Key: "$set", Value: bson.D{
		{Key: "meta", Value: bson.D{
			{Key: "$arrayElemAt", Value: bson.A{"$meta", 0}},
		}},
	}}}

	// Runs the pipeline
	cursor, err := collection.Aggregate(context.TODO(), mongo.Pipeline{searchStage, facetStage, setStage})
	if err != nil {
		panic(err)
	}

	// Prints the results
	var results []bson.D
	if err = cursor.All(context.TODO(), &results); err != nil {
		panic(err)
	}
	for _, result := range results {
		fmt.Println(result)
	}
}
